### PR TITLE
Feature/#531/more powerful list command

### PR DESCRIPTION
**Pull request by ElijahSink**
## Pull Request Check List

- [x] A news fragment is added in `news/` describing what is new.
- [x] Test cases added for changed code.

## Describe what you have changed in this PR.
Added a new mode `--json` to the `list` command which outputs the dependency graph as a JSON document.

### CODE REVIEW

File: pdm/cli/utils.py [hunk 1]
```diff
@@ -291,23 +301,107 @@ def _format_forward_dependency_graph(project: Project, graph: DirectedGraph) ->
...
+        top_level_packages = filter(
+            lambda n: not list(graph.iter_children(n)), graph
+        )  # leaf nodes
+        build_dependency_json_subtree: Callable = build_reverse_dependency_json_subtree
+    else:
+        top_level_packages = graph.iter_children(None)  # root nodes
+        build_dependency_json_subtree = build_forward_dependency_json_subtree
+    return [
+        build_dependency_json_subtree(p, project, graph)
+        for p in iter(sorted(top_level_packages, key=lambda p: p.name if p else ""))
```
Comment (frostming) [inline pdm/cli/utils.py:388]: `iter()` is redundant?

**Review by frostming — COMMENTED** [review]
Be careful about circular dependencies, which may cause an infinite loop. You may also need to include such in the test case

File: pdm/cli/utils.py (hunk 1 above)
Comment (ElijahSink) [reply pdm/cli/utils.py]: Yep, thanks! Removed.

**Comment by ElijahSink** [comment]
The existing test case for circular dependencies results in no output for both the normal-reversed and json-reversed list actions. This is because the simple test case (direct circular dependency between `foo` and `foo-bar`) has no leaf nodes, which are our starting points for listing in reverse.

This being the case, I've modified the test cases for reverse listing a circular graph. Perhaps there's a better way to find "pseudo-leaf nodes" in a purely-circular dependency graph? Perhaps it doesn't matter.

**Comment by frostming** [comment]
Good enough, let's see how to improve the rest in another go. Thank you for taking care of this!

**Pull request merged**